### DISCUSSION TITLE
Thread: Mistakes in face clustering, Shit clusters, Faces not showing up

### DISCUSSION THREAD

**Discussion by IndrekHaav** [nextcloud/recognize#754] [General]
I have about 200 photos uploaded to my Nextcloud instance, most of those containing faces. Recognize has processed them all, but only a single person shows up in the People section (of both Photos and Memories apps), with 12 photos.

I don't think this is the same as #588. I have checked the database, and the `oc_recognize_face_detections` table has 237 records, while `oc_recognize_face_clusters` only has one. Furthermore, if I manually insert a record into the clusters table and then link a record in the detections table to it, it shows up as a new person. I don't understand why the majority of the detected faces are not added to a cluster.

The same photos, when imported into PhotoPrism (which also uses Tensorflow) resulted in every single detected face showing up.

**Comment by marcelklehr** [comment]
That would indicate problems with the clustering algorithm. It sounds like it doesn't run at all, though, because it should create at least some clusters, IMO.

**Comment by IndrekHaav** [comment]
Like I said, it did create one cluster, with 12 faces. Those were all the same person, too, so what work it did, it did correctly.

But just to clarify, the clustering algorithm should be able to create clusters with even a single face? There's no minimum number of faces that need to match before a cluster is created?

Edit: is there any way to run just the clustering job from the command line? With verbosity turned up?

**Comment by marcelklehr** [comment]
> here's no minimum number of faces that need to match before a cluster is created?

There is a minimum. It's currently at 6 detections.

> is there any way to run just the clustering job from the command line?

not at the moment, you can manually create an entry in oc_jobs in your database, though, for \OCA\Recgonize\BackgroundJobs\ClusterFacesJob with argument `{"userId": "<your user name here>"}`

**Comment by IndrekHaav** [comment]
> There is a minimum. It's currently at 6 detections.

I see. That might explain why faces from my photos aren't showing up. Can I ask why this minimum? And is there any way to change it?

**Comment by marcelklehr** [comment]
@IndrekHaav The clustering algorithm needs a few hyperparameters to work well. From testing it transpired that setting a minimum cluster size improves clustering because it prevents accidental face matches to agglomerate into larger clusters that don't represent a single person (Something I like to call *shit clusters*). I recommend trying out v3.5.0 first to see if that improves the situation for you, since we're shipping a new clustering algorithm with that release. If that doesn't help and you're adventurous, you can change the min cluster size constant here:

https://github.com/nextcloud/recognize/blob/015f3c888af7f002b199c2414489769294a5fde6/lib/Service/FaceClusterAnalyzer.php#L20

(We've reduced the value from 6 to 5 in v3.5.0 now)

In v3.5.0 there are now convenience occ command for resetting clustering and running clustering manually: `occ recognize:reset-face-clusters` and `occ recognize:cluster-faces`

**Comment by IndrekHaav** [comment]
Thanks for the response!

I tried the new algorithm in 3.5.0. Just to be safe, I wiped all detected faces and clusters from the DB and triggered a full re-crawl.

This time, it created a few more people, but the vast majority of faces were put into a single cluster, seemingly almost randomly. I retried this a few times and also reran the `recognize:cluster-faces` command a couple of times; the exact numbers varied a bit, but I always ended up with 5-6 clusters, one of them with 80-90 photos of at least a dozen different people (male and female, from infants to 80-year-olds).

In other words, a *shit cluster*.

Having incorrect faces in the cluster wouldn't be so bad as they can be removed in the UI, but the problem is that there's no way (at least that I could see, in the Photos or Memories apps) to move photos from one person to a new person. One can only move them to an existing person, or remove them from the cluster completely. I tried the latter, but subsequent `recognize:cluster-faces` commands only put them back (I was hoping they would at least be put into new clusters).

I tried changing `MIN_CLUSTER_SIZE`, and found that lower values seemed to work better. I set it to 2, and that produced 13 clusters, with overall much better distribution of faces between clusters. That makes sense to me - some people appear in only very few photos, and a higher minimum cluster size might force the algorithm to incorrectly combine them with other faces. I tried setting it to 1, but that produced an error when running the clustering command. I would have liked to see the result, though - even if it errs too far in the other direction, I would rather have more clusters that I have to merge, than fewer clusters that I cannot effectively split.

While I'm messing with the code, is there another constant or parameter that determines how similar the faces have to be to get clustered together?

**Comment by marcelklehr** [comment]
> This time, it created a few more people, but the vast majority of faces were put into a single cluster, seemingly almost randomly.

How many images do you have?

>  tried the latter, but subsequent recognize:cluster-faces commands only put them back

that would be a bug

> I would rather have more clusters that I have to merge

yeah, that makes sense

> While I'm messing with the code, is there another constant or parameter that determines how similar the faces have to be to get clustered together?

There is no constant value that governs this. HDBSCAN is an adaptive algorithm that learns from the density patterns of the data. The more files you have, the better the outcome. You could also try playing with MIN_SAMPLE_SIZE.

**Comment by EVOTk** [comment]
Hello,
I don't know if this has anything to do with this ticket, but since the update to 3.5.0

I don't have any "peoples" in Photos. and the `occ recognize:cluster-faces` command is broken:
![image](https://user-images.githubusercontent.com/45015615/218152562-53145dec-1f21-488e-a4d5-75097caa7647.png)

I had more than 2500 detections

> OMV6
> Docker linuxserver/docker-nextcloud 
> Nextcloud 25.0.3
> Php 8.0.25

Oh, it looks more like this problem https://github.com/nextcloud/recognize/issues/676

**Comment by IndrekHaav** [comment]
> How many images do you have?

229. This is just a test set, I'm evaluating Memories + Recognize as a replacement for Photoprism.

I did try setting `MIN_SAMPLE_SIZE` to 2 as well. That also seemed to improve things a bit - created a few more clusters, and incorrectly grouped faces were at least more reasonable (visually similar, same gender and age group, etc).

**Comment by MB-Finski** [comment]
> > How many images do you have?
> 
>     229. This is just a test set, I'm evaluating Memories + Recognize as a replacement for Photoprism.
> 
> 
> I did try setting `MIN_SAMPLE_SIZE` to 2 as well. That also seemed to improve things a bit - created a few more clusters, and incorrectly grouped faces were at least more reasonable (visually similar, same gender and age group, etc).

Like @marcelklehr already commented, HDBSCAN, as it is implemented, will try to find the most stable clusters from the data regardless of their size. If the data contains a large number of identities (especially multiple identities with fewer face detections than MIN_CLUSTER_SIZE) it'll still try to find the most stable clusters in the data. This can lead to combining multiple identities of similar looking persons. The easiest way to alleviate this issue is to scan in a larger dataset.

Mind you, "similar looking" to the face recognition model may not always be similar looking to you or me. This is especially true in the case of children/infants. The dlib face recognition model hasn't been trained with images of children so they will cause trouble with clustering regardless of the clustering algorithm. (IIRC, Photoprism, for example, had a community effort to retrain their recognition model with datasets containing images of children.)

MIN_SAMPLE_SIZE is basically a probability density (i.e. "face detection density") smoothing factor used by HDBSCAN. Reducing this value *too much* can lead to statistical noise causing issues with larger datasets. Still, it might be that we'll have to reduce this value going forward; the optimal value in my test dataset may not be optimal for all users. Also, the optimal value will depend, to some degree, on how incremental clustering is implemented as that affects the amount of noise in the data that is being clustered.

@marcelklehr : Besides fine tuning MIN_SAMPLE_SIZE, another way to improve the clustering in this case might be to implement a limit on the maximum size of a cluster. The obvious, but likely(?) not optimal, solution would be to limit the radius of a face cluster. However, limiting the maximum edge length within a cluster (this was implemented in a previous version of the MstClusterer-class but I stripped it since it was not used) may be a better solution since this will specifically limit forming clusters in sparse areas of the face embedding space (since the mutual reachability distance will also be large in these areas). If the latter is implemented, it may help us get away with a larger MIN_SAMPLE_SIZE which is better for users with larger datasets. It may also be that a combination of both of these limits would provide the best user experience.

**Comment by IndrekHaav** [comment]
@MB-Finski Thanks for the info, that's an interesting read!

However, coming back to the original issue - irrespective of the way the clustering algorithm works, I think there should be a way for the user to review recognised but unclusters faces and, for each one, choose between "not a person" (don't suggest again), "merge with ______" (pick existing cluster) or "new person" (create new cluster).

Or is that something that should be handled by another app like Photos or Memories?

**Reply by marcelklehr** [reply]
I think this is a good long term plan! The UI will need to be implemented in photos/memories, but recognize will need to provide the backend.

**Comment by ced455** [comment]
I agree with @IndrekHaav, be able to create a new person is missing when trying to filter out false positive.
I am in the situation where i need to remove face from a person and i cannot find it back in the people tab once done, is this face lost forever ?

**Comment by marcelklehr** [comment]
> remove them from the cluster completely. I tried the latter, but subsequent recognize:cluster-faces commands only put them back (I was hoping they would at least be put into new clusters).

I cannot reproduce this. For me removed face detections are not readded to the same cluster anymore.

**Comment by IndrekHaav** [comment]
@marcelklehr For me, every time the clustering job ran, the same faces kept getting added to the same person. I ended up deleting the detected faces from the DB, because they were faces I wasn't interested in anyway (random background people, and such).

How would this work anyway? Does the app keep track of clusters that a face has been removed from in some way?

**Comment by marcelklehr** [comment]
>  limiting the maximum edge length within a cluster [...] may be a better solution since this will specifically limit forming clusters in sparse areas of the face embedding space

@MB-Finski If you're up for implementing that, I'm happy to merge a pull request (let me know if you need help with git).

**Comment by marcelklehr** [comment]
> How would this work anyway?

When removing a face from a cluster we store the distance from the cluster centroid along with the face and in the future only add it to a cluster if the distance to the cluster centroid is smaller.

**Reply by Olen** [reply]
Where is this information stored?
Is everything stored in the database, or is there some "AI-magic" stored elsewhere?
I have spent quite a few hours moving photos from one person to another and cleaning up the clusters using "Memories", but obviously only for my own user.
Now I want to replicate the results for other users (shared groupfolders), without repeating all the work.
I can easily run a database-query to change the cluster_id based on file_id, x, y and vector, but if "Move to another person" in the GUI does more than that, I would need to replicate that as well, and that might not be as easy.

**Reply by marcelklehr** [reply]
The min-distance from cluster centroid is stored in the database for each detection.

**Reply by Olen** [reply]
Thanks for the quick reply.  Does this mean that doing a simple `update oc_recognize_face_detections set cluster_id=123 were cluster_id=321` is all it takes?
Or should another process be run as well to calculate and update the new min-distance?
And can/will this process be run automatically/periodically?

**Reply by marcelklehr** [reply]
Yes, that's all it takes

**Comment by marcelklehr** [comment]
@MB-Finski I wonder if it would make sense to fall back to DBSCAN clustering for photo collections smaller than x photos, as HDBSCAN results are pretty wild on smaller collections.

**Comment by rhatguy** [comment]
Coming over to this discussion thread for this since #721  was closed.  I'm still experiencing this with V3.7.0.  Should we be waiting for the pull in #711 to be released before doing more testing?

**Reply by marcelklehr** [reply]
#711 is part of v2.7.0 :/ 

Did you recalculate the clusters?

**Reply by rhatguy** [reply]
![recognize](https://user-images.githubusercontent.com/11964411/227563708-ab9e218f-3ee4-4ebc-b71d-eb07e68e95d2.JPG)

  - memories: 4.12.4
  - recognize: 3.7.0

This is running with the standard Nextcloud docker container fully up to date running under Unraid on a large X86 server.  Not in WASM mode, and not using a GPU.

I ran all of the resets available via occ, then ran a new classify job.  Note in the attached picture that 3 different people have been clustered as the same person.  There are many other examples of mis-clustered people within this same cluster as I just took an example all from the same day.  It seems like the clustering is missing badly.

**Reply by rhatguy** [reply]
I installed the newest 4.0.1 version of recognize, ran all the resets again and re-ran classification then clustering.  This same group of people as I pasted the screenshot above got clustered together again (along with many other clusters that aren't even close.  It seems like the clustering algorithm needs to be completely re-worked to be useful.

**Reply by marcelklehr** [reply]
The faces in your screenshot are all shot from the side, which probably causes issues in recognizing them. Also, are you using recognize in WASM mode or native?

**Reply by rhatguy** [reply]
Native mode.  Running the standard Nextcloud docker container on a large Unraid server.

Any chance in recognize leveraging xmp face names in existing metadata?  If recognize could read existing names from metadata and assume those names to be correct those of us who have spent time curating that data would already have a large number of verified clusters.  In a thread on the Memories GitHub it was suggested that these tags should be read by recognize and loaded into the existing recognize tables even if recognize doesn't actually use the data for clustering new detections.  If the data could be used by recognize clustering algorithms it would give people an option to hint or correct mis-clustered detections once and for all as recognize detection/clustering is rerun.

**Reply by marcelklehr** [reply]
> Any chance in recognize leveraging xmp face names in existing metadata?

There is an open issue for this (#761) but Nextcloud GmbH is not looking to sponsor feature development on recognize at the moment (see https://github.com/nextcloud/recognize/discussions/779). I might get to working on this in my free time, though. In the meantime any contributions are of course welcome :)

> Native mode. Running the standard Nextcloud docker container on a large Unraid server.

*scratches head* standard Nextcloud docker container doesn't support libtensorflow AFAIK. Libtensorflow relies on glibc, but the nextcloud container uses musl. You may be running in WASM mode without knowing.

**Reply by rhatguy** [reply]
Doesn't the following in the administration panel mean WASM mode is not enabled?  I'm running the docker image pulled from nextcloud:latest under unraid.

root@Tower:~# docker inspect Nextcloud | egrep -i image
        "Image": "sha256:3a5ef33e5056185c95d428e234c2a193844650b417dfa82984786de9a5ee54de",
            "Image": "nextcloud:latest",


![recognize_tensowflow](https://github.com/nextcloud/recognize/assets/11964411/6b3fc694-e41a-486c-a802-fb4f7cacd9ca)

**Reply by marcelklehr** [reply]
Ah, yes, that one only checks for CPU compatibility. Can you post the check at the bottom for node.js?

**Reply by rhatguy** [reply]
![recognize_tensowflow](https://github.com/nextcloud/recognize/assets/11964411/607b480c-6f30-441d-b6b5-66f499b10cdd)

**Reply by marcelklehr** [reply]
huh. Did you do anything special for this to work, or did it just work out of the box?

**Reply by marcelklehr** [reply]
Ah, right. The default container is debian, not alpine. Got that wrong.

**Reply by rhatguy** [reply]
Just confirming...nothing special was done.  I've tried to keep my Nextcloud installation as vanilla as possible (other than installing apps inside Nextcloud) to keep it easy to upgrade/maintain.

**Comment by MahdiBaghbani** [comment]
I have just installed Recognize in Nextcloud and uploaded 100 purely portrait photos, while Recognize tags all them as portraits, people as it should, it fails to find a single face in them.

Face score too low. continuing with next face.
Face is not straight. continuing with next face.

**Reply by marcelklehr** [reply]
100 photos may not be enough. The clustering algorithm needs a minimum of 100-200 faces to generate useful clusters.

**Comment by marlemion** [comment]
With recent nextcloud docker container (26), face recognition/clustering must be classified as not working at all. While clusters are created, people are added completely randomly. I wonder whether this feature ever has been tested on a real-life set of pictures? I have about ~30 000 pictures, of which approximately at least 15% have persons on them (I use to be in nature a lot). However, the result is so discouraging that I have not even thought of manually reordering persons and pictures. I remember having set up the dedicated app facerecognition at some point with nextlcoud 24 I think, and that worked way better, but it required a special docker image to be used.

**Reply by jmigual** [reply]
Indeed, I'm having the same issue. It seems that letting it run longer and feeding it more data works better but not that much better... The clustering algorithm doesn't work that well at all unfortunately...

**Reply by marcelklehr** [reply]
> I wonder whether this feature ever has been tested on a real-life set of pictures?

It has :/

I wonder, are you using WASM mode or normal tensorflow mode?

**Reply by marlemion** [reply]
I am not using WASM (I am on the official docker image):

![grafik](https://github.com/nextcloud/recognize/assets/5897656/dab8c97c-1f5f-4a95-bcb4-901b29fed5a1)

**Comment by MB-Finski** [comment]
Sorry about this taking a while; I've been too busy at work. Anyhow...

So, just a quick recap of what I think is going on here (just in case someone else has time to work on this):

1. Although the underlying HDBSCAN is state-of-the-art (qualitywise) in density based clustering, HDBSCAN is not inherently suitable for incremental processing of embeddings. This means some additional algorithmic complexity (and failure points) when reconciling the results of consecutive clusterings with existing clusters. While the current solution can surely be improved with tweaks in the current code base, ultimately better results could be had by swithing HDBSCAN to FISHDBC. FISHDBC is essentially an approximation of HDBSCAN but with support for incremental processing. FISHDBC would also be computationally much more efficient than the current solution. (Not to mention that efficient large scale clustering would have various applications when working with LLMs).

2. The hyper parameters for the face clustering are currently (likely) far from ideal and tuning them so that everyone gets the best possible results is quite tedious. If you find parameters that work for you, please report back!

3. As discussed above the face recognition deep neural network currently in use (dlib) is not too good with side profile pictures (or children).  The density/radius of these "shit clusters" is identical to proper face clusters so from the viewpoint of the clustering algorithm there's no practical way of knowing which is which. At some point we'll probably want to upgrade to a model that has been trained with side profile pictures. Some face recognition model architectures (arc-face) even output a meaningful measure of how sure the model is with the calculated embedding which might be useful.

-----------------------------------------------------------------------------------------------------

I've made a fork that has (for my personal library) vastly improved performance (mostly improving on points 1 and 2 from above). I'm still running into issues when getting over 30k face detections/user due to the batch prosessing not working correctly in NC 26 but the clustering results are much better for all my users. I intentionally set the hyperparameter such that splitting of identities is more likely than fusing them since this, I think, is the preferred type of error instead of fusing identities.

If you try this fork, please, report back with your results. Only one file needs updating: https://github.com/MB-Finski/recognize/blob/main/lib/Service/FaceClusterAnalyzer.php

**Reply by rhatguy** [reply]
I did a test using the fork today by replacing the php file in my docker install.  Unfortunately its difficult to comment on how much "better" the fork is.  Clustering does seem to work pretty good, when its working.  However pretty early in recognize's run through my picture collection I get a crappy cluster created with the pictures I've pasted earlier in this thread.  It seems like once that cluster gets created, it ends up gobbling up random other people who clearly don't belong in that cluster.  In clusters that haven't turned to crap, the clustering looks quite good (with the fork).  Maybe it is those side profile faces that are blowing out the parameters and once the cluster starts matching random people its creates a black hole effect.   It seems like if there were a way to constrain the size of the cluster, or cause recognize to stop using those offending pictures for further clustering that would be helpful.  I'll continue to let this run and report back further on the entire image set.

**Reply by Andreaux** [reply]
Indeed, a "Remove from person and don't match anymore" flag would be a nice addition I guess. But then also a way to list "banned" faces so you can reverse that decision later would also be needed :)

**Reply by Andreaux** [reply]
Maybe simply a "Don't match face" flag and flagging a face would remove it from clusters.

**Reply by MB-Finski** [reply]
@rhatguy 

Thanks for reporting back! It sounds like it is the incremental prosessing that ruins the clustering for you, then. Now that we have a bit more suitable hyper parameters, you should get better results by removing the clusters and running the clustering from cli once you've scanned in all faces (this results in a single HDBSCAN run that clusters all faces at once). Unfortunately, it is inevitable that one or more of these "shit clusters" form even during full HDBSCAN runs (until we switch to some other recognition model). However, these clusters should not consume other identities. If they do, then this suggests that we may need a bit more conservative clustering hyper parameters. 

With incremental processing we also have the issue that the noise (clustering rejects) accumulates during successive runs while the number of valid face detections decreases (we only sample previously identified clusters so as to reduce the computational cost). @marcelklehr has an improved version of this fork in the works. This new version also samples the the clustering rejects which should keep the signal to noise -ratio, so to speak, more consistent. This hopefully should reduce the risk of both shit clusters forming and successive runs gobbling up random identities.

EDIT: The flagging of bad clusters/face detections like what you both have suggested is a good idea, IMO.

**Reply by Andreaux** [reply]
Honestly I feel "shit" clusters are less of a problem. My issue is more with faces straight not getting in any clusters at all and photos with more than one person being attributed to only one of the faces in the picture, only once. I might be wrong but I feel we're missing a lot of faces like that. If a face is in, but in the wrong cluster, you can at least move it, but if it never get in in the first place, that's more of an issue. Honestly, even a cluster with 1 or 2 faces is more useful than leaving it out. To me at least :)

**Reply by marcelklehr** [reply]
> Honestly, even a cluster with 1 or 2 faces is more useful than leaving it out.

The problem is that if we allow a cluster with 1 or 2 faces already it's likely to become a shit cluster.

> photos with more than one person being attributed to only one of the faces in the picture

Each photo can only be added to a cluster once, but I cannot reproduce one photo with multiple faces only getting assigned to one cluster.

**Comment by Andreaux** [comment]
On this topic, I was wondering how Recognize deals with pictures with more than one person in. I might be wrong, but it seems to me as **once a picture has been attributed to a person, it is not evaluated for further persons in the image**. Could it be? I have a lot of pictures with more people in and once one face is recognized and the picture shows up in that person's "set", the other persons don't seem to show up anywhere, even when there's plenty of pictures with them and their face is clearly visible (even maybe better than the one originally picked). This makes me lose quite a bunch of people.

**Reply by marcelklehr** [reply]
As mentioned in the other thread: Each photo can only be added to a cluster once, but I cannot reproduce one photo with multiple faces only getting assigned to one cluster.

**Reply by Andreaux** [reply]
For example, I have a lot of photos of my son and daughter in the same picture and instead of having all the pictures under both of them, part of the pictures end up with one of them, while the other part with the other… I have reset the whole a couple of times and got similar results. Same with colleagues, etc.

**Reply by Andreaux** [reply]
I also gave Finski’s version a shot and while I got slightly different results, the 1 photo 1 face behavior seems the same…

**Reply by Andreaux** [reply]
I’ll make explicit tests about this today and get back to you. I can even provide test images, maybe we can better pinpoint what is going on…

**Reply by marcelklehr** [reply]
> I can even provide test images

That might be helpful!

> I have a lot of photos of my son and daughter in the same picture

For children this might happen because the model embeds children's faces rather close to each other so that both children in the photo will get assigned to the same cluster, which then removes one of them.

**Reply by marcelklehr** [reply]
> which then removes one of them

Come to think about it, we can actually get rid of that now.

**Reply by Andreaux** [reply]
> For children this might happen because the model embeds children's faces rather close to each other so that both children in the photo will get assigned to the same cluster, which then removes one of them.

I get that clustering kids is a challenge (also grouping images from the same kid at different ages, when they have changed a lot in a short time), but I observed the same with adults too.

**Reply by Andreaux** [reply]
I just did some tests and have to conclude that, in some cases, - you're right - photos with more than 1 recognizable person in are indeed attributed to both persons, but there are definitely cases where this doesn't happen. I'm not sure why though.
I have specific examples (only kids and mixed) where it doesn't happen. I have yet to find examples where all persons are adults and the amount of overlapping images would be high enough to justify a face to be created.
Here's an example:
![Screenshot 2023-05-17 at 15 54 27](https://github.com/nextcloud/recognize/assets/1387888/376e1ede-3ddf-4b4c-8e99-f5999fc88ffe)

**Reply by marcelklehr** [reply]
I think these may be filtered out because the face is rotated too much on one or multiple axes.

**Comment by marcelklehr** [comment]
I have some news after running a benchmark on the (now improved) clustering algorithm using a part of the IMDb-Face dataset.

Here are the numbers:

### Detected faces out of all faces
86% (possibly overestimated)

### Clustered faces out of all detected faces
28%

### Clustered faces out of detected faces per identity
59%

### Identities with clusters out of all identities
87%

### Shit clusters out of all clusters
2%

### Correctly assigned faces out of all assigned faces per cluster
92%

Note, that the code that these numbers are based on has not been released yet.

**Reply by Andreaux** [reply]
Wow! Sounds promising!

**Reply by marcelklehr** [reply]
> Clustered faces out of all detected faces
> 28%

I'm not sure how we can improve this

**Reply by marcelklehr** [reply]
The new code is now available as part of v4.1.0 for Nextcloud 26+

**Reply by ced455** [reply]
Thank you @marcelklehr for all the hard work ! Should users just reset cluster with `occ recognize:reset-face-clusters` and then run `occ recognize:cluster-faces` to take advantage of 4.1.0 or shoud we reset everything ?

**Reply by marcelklehr** [reply]
`occ recognize:reset-face-clusters && occ recognize:cluster-faces` should be enough. Note that cluster-faces by default will cluster all faces at once, which might be too much for your RAM, so I recommend doing it in batches of e.g. 10000: `occ recognize:cluster-faces -b 10000`

**Reply by rhatguy** [reply]
Quick update from 4.1.0.  I reset all clusters and re-ran a clustering job across my 41k pictures (~30k faces) in a single clustering run.  Without going to in depth the clustering seems MUCH better now.  There are still some mistakes, for instance, many pictures of my son and daughter got clustered together and I have one cluster which seems to have picked up random people.  I'll have to dig in a bit more to see if there are any recognizable patterns that additional tuning might be able to fix.  Nice work though!  Recognize seems to be approaching the clustering accuracy that I see from Digikam with this update.

**Comment by Andreaux** [comment]
I ran into something strange today. I see a different amount of faces in Photos/People and Memories/People. Aside the fact that some faces are blank in Photos and don't open any photos when clicked, I see 1 less persons in Memories. The difference is a person with only 1 photo and that photo is in a hidden folder outside of /Photos. Is there a setting to limit the scope of Recognize to not go out of the Photos (or whatever is set to hold pictures) folder AND not go into hidden folders?

**Reply by marcelklehr** [reply]
Recognize listens to `.nomedia` files. Put one in the folder you want it to ignore

**Reply by Andreaux** [reply]
Not ideal, each Photo application lets you choose a root folder, which is set to /Photos for me. Funnily, if I put a .nomedia file in the other folders outside of Photos, I get errors like "Failed to load some photos" or "No valid root folder found (.nomedia?)" from other apps like Memories (which is instructed to look in /Photos only). I'll leave it like that with the .nomedia files and see on the long term.

**Reply by marcelklehr** [reply]
You may need to rescan in memories.

**Reply by Andreaux** [reply]
Probably. Have yet to find how it can be done :)

**Comment by iamdoubz** [comment]
I have a question, I installed Memories and Recognizer. Memories reports that 85,352 media files have been indexed. I then ran a full classification manually `occ recognize:classify`. It took approximately 36 hours (for others who are wondering). Now, I am in the Memories app and click on People. My face shows up at least 8 times. My question: how do I combine these into one?

**Reply by jmigual** [reply]
From memories, you can go inside each face and on the top right menu select, "Merge with other"  (or something similar, I'm saying this from memory). You can also move individual photos to another face if they are misclassified by selecting them and going to the top right menu

**Reply by iamdoubz** [reply]
Thank you for replying! I couldn't find this option in Memories (I think) but did find it within Photos app.

Now better/harder question: does this act retrain the model in Recognizer?

**Reply by jmigual** [reply]
Kind of but not completely. Each face has a numeric value representing it and the clustering algorithm creates clusters based on the similarity of this numeric value. When you move faces together it takes an "average" of the cluster to decide the value of the face. The more photos you have the more accurate the "average" is to the reality of the face and may sometimes perform better. That is in theory but in practice I've seen very little improvement.

**Reply by iamdoubz** [reply]
Thank you for the detailed explanation. Theory versus actual is always a fun guessing game. It was just interesting that pictures taken at the same time would have two different clusters for virtually the same pictures. 

Last two things then I actually have to start my day job: 
- "A minimum of 120 faces per user is necessary for clustering to kick in" It says that I have 31 faces left to process but the background job won't trigger because it needs 120 photos? What command `occ` can I run to just say "do it anyway"?
- Consequently, when I run an `occ recognize:bla` command, can I pass a parameter to only use X threads? First time I ran the whole classify item it used all my cores and my server was unusable until it completed and I do not want that to happen again...

**Reply by marcelklehr** [reply]
> What command occ can I run to just say "do it anyway"?

occ recognize:cluster-faces

(You will likely want to use the batch size parameter for this, to avoid blowing up your RAM)

> when I run an occ recognize:bla command, can I pass a parameter to only use X threads?

No, there's no cli param for number of cores. You will need to adjust this in the admin settings UI.

**Comment by luxzg** [comment]
I have a question, and all I've read leads nowhere or here :)

Been using recognize since about March 2023.

Over time my collection of digital photos from about 2007-2024 keeps on growing, probably have few hundred thousand photos, 

As a result I have several big clusters, like me and close family, at about 1200-1300 photos in each. But I also have 11927 "unassigned" faces. And all know clusteres still total way under 10k, so unassigned outnumber all "assigned" ones. To add to that, clustered number includes a big shit-cluster of around 1.1k photos which I named Trash and throw innocent bystanders into it (for lack of better management).

What amazed me that my iwn face and few members of the family simply stopped being clustered. If I open unassigned group, we're all there, probably 1-2k photos (or faces) of us just from this summer. But nothing gets clustered.

And yes, cron jobs are running, and after jumping from NC26 to NC29 I kept trying manually running scans, classifications, clustering and so on. Yet the unassigned cluster keeps growing, and new additions to any actual person cluster stays very low. It is not zero, but there's like under 100 random photos in total added in past ~18 months to all named clusters in total. My own face zero, as well as two other clusters with 1k+ tagged photos.

So here I am wondering - can I tweak something? Maybe something hardcoded in PHP? Like a threshold when a face gets recognized as part of cluster? I'd rather have a bit more false positives than not grt anything clustered anymore.

Oh and btw, I did try manually moving unassigned to correct persons, but neither Photos nor Memories allow that. So that doubles my woes, as I can delete or remove wrong tags (eg a stranger being tagged as me), and I'm willing to do that cleanup manually from time to time, but currently I can't in any way force hundreds of photos of myself to get moved to my personal cluster. (Using myself as example, same goes for other family members).

I'd be happy to test any ideas. Thanks in advance!

**Reply by iamdoubz** [reply]
When you are in the "Unassigned faces" section, you can select one or more photos and click move to another person. That brings up another window asking you what "face" AKA person to match the selected photo(s) to.

![image](https://github.com/user-attachments/assets/54bed43f-5f34-4c8e-ae20-1cc0ea6c29dc)

This is the manual method that I have done anyway.

**Reply by luxzg** [reply]
> When you are in the "Unassigned faces" section, you can select one or more photos and click move to another person. That brings up another window asking you what "face" AKA person to match the selected photo(s) to.
> 
> This is the manual method that I have done anyway.

OMG thanks for the picture, otherwise I'd claim it wasn't there! I completely disregarded that first icon, I went (multiple times!!) to the three-dot "hamburger" menu and it wasn't there. Zeah, call me old fashioned, but I'd expect expanded menu to contain ALL options, not "hide" one by just showing icon in the pop-up. But yeah, I tried with one phozo, and it works, that will be workaround enough for now. Still, I'd preffer automatic clustering to do better job, as it did in some previous version (IDK which). But thanks a lot for shoving this into my face so I notice 👍🏼 👍🏼

**Reply by iamdoubz** [reply]
Don't worry, someone had to show me too so you aren't alone! Just paying it forward.

**Reply by Andreaux** [reply]
> OMG thanks for the picture, otherwise I'd claim it wasn't there! I completely disregarded that first icon, I went (multiple times!!) to the three-dot "hamburger" menu and it wasn't there. Zeah, call me old fashioned, but I'd expect expanded menu to contain ALL options, not "hide" one by just showing icon in the pop-up. 

This is not being old-fashioned, that's a UX fail. When someone doesn't find something and it turns out that others reply "don't worry, someone had to show me too", this is the sign of some kind of UX/UI failure. Indeed, all options have to be contained in a menu and icons outside of the menu can only serve as shortcuts. Luckily it's not that bad with this one, here I'd rather blame the icon. It should show a head and an arrow pointing towards the head, that would help a lot. This should be corrected IMHO as apparently a lot of people run into this issue.
Same goes for the "unknown faces" button... it is so not obviously placed :)
I'm grateful for it being there, but damn is it not obvious...

**Reply by marcelklehr** [reply]
Feel free to open an issue in the memories repo.

**Reply by luxzg** [reply]
> Feel free to open an issue in the memories repo.

It's already there:
https://github.com/pulsejet/memories/issues/1341

Edit: though it's tagged as feature request, I'm unsure if this should be a bug or a feature...

Edit 2: and while this workaround through Memories app exists, it would still be nice to find a way for recognize to better cluster existing photos. @marcelklehr could you chime in on this? Is there any number that can be tweaked on my own somewhere to allow for a more "loose" clustering, sure it may cause some false positives, but I'm prepared to try tweaking any variables until I find values that work better for me

**Reply by marcelklehr** [reply]
> find a way for recognize to better cluster existing photos

Totally agreed! Recognize's tech stack is somewhat outdated at this time, so we're planning to implement a revamp some time soon. If you want to tinker a bit with the parameters until then, here's the parameters for the clustering algorithm: https://github.com/nextcloud/recognize/blob/main/lib/Service/FaceClusterAnalyzer.php#L18-L24

**Reply by luxzg** [reply]
> > find a way for recognize to better cluster existing photos
> 
> Totally agreed! Recognize's tech stack is somewhat outdated at this time, so we're planning to implement a revamp some time soon. If you want to tinker a bit with the parameters until then, here's the parameters for the clustering algorithm: https://github.com/nextcloud/recognize/blob/main/lib/Service/FaceClusterAnalyzer.php#L18-L24

Awesome, I van work with that! 😃 And read about the revamp on another issue as well, looking forward to it, it's probably my favorite Nextcloud feature (speaking as a home/family user). Thanks for all you people do on these projects!
